anesthetic/weighted_pandas.py

```diff
@@ -1,14 +1,67 @@
 """Pandas DataFrame and Series with weighted samples."""
 
+import warnings
 from inspect import signature
 import numpy as np
 from pandas import Series, DataFrame, concat, MultiIndex
+from pandas.core.groupby import GroupBy, SeriesGroupBy, DataFrameGroupBy, ops
+from pandas._libs import lib
+from pandas._libs.lib import no_default
+from pandas.util._exceptions import find_stack_level
 from pandas.util import hash_pandas_object
 from numpy.ma import masked_array
 from anesthetic.utils import (compress_weights, channel_capacity, quantile,
                               temporary_seed, adjust_docstrings)
 
 
+class WeightedGroupBy(GroupBy):
+    """Weighted version of ``pandas.core.groupby.GroupBy``."""
+
+    grouper: ops.BaseGrouper
+    """:meta private:"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def mean(self, *args, **kwargs):  # noqa: D102
+        result = self.agg(lambda df: self.obj._constructor(df).mean(
+            *args, **kwargs))
+        return result.__finalize__(self.obj, method="groupby")
+
+    def std(self, *args, **kwargs):  # noqa: D102
+        result = self.agg(lambda df: self.obj._constructor(df).std(
+            *args, **kwargs))
+        return result.__finalize__(self.obj, method="groupby")
```

Comment:
I agree that it's far from ideal, but I don't see the alternative as being maintainable long-term as new arguments are added in future versions of pandas. In principle, if desired, we could implement a sensible ddof, but I'm only inclined to do that if there is a clear use case.

Reply:
I would much rather have the explicit args and kwargs, and _that_ would actually be the pythonic way, but we can't maintain a full copy of pandas and keep it up to date with all the little changes going on there. Hence it would have been much nicer to have GroupBy automatically point to the corresponding methods, without needing to create a GroupBy method for every WeightedDataFrame method that we want to use...

For any anesthetic-own functions I would strive to be explicit, but when it comes to inheritance from pandas, I would try to keep things as independent and robust to changes in pandas as possible.

Reply:
Specifically `ddof` is probably something we could improve upon. `ddof` actually _does_ make sense for weighted data. For MCMC data (with integer weights, aka repeat type weights, aka frequency type weights) you can use it actually one to one compared to unweighted data. For nested samples (which I guess would fall into the category of reliability type weights) I am not sure what the best way of handling this would be, but just dividing by the sum of weights does not give you an unbiased estimator...

See the corresponding [Wikipedia article](https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Weighted_sample_variance) for an overview. But note that the section on reliability type weights might not be correct, see also this [Stackoverflow answer](https://stats.stackexchange.com/a/61641/209204).